hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/PreHandleWorkflowImpl.java

```diff
@@ -54,192 +57,208 @@
+        this.signatureVerifier = requireNonNull(signatureVerifier);
     }
 
+    /** {@inheritDoc} */
     @Override
-    public void start(@NonNull final HederaState state, @NonNull final Event event) {
-        preHandle(requireNonNull(event).transactionIterator(), requireNonNull(state));
-    }
+    public void preHandle(
+            @NonNull final ReadableStoreFactory readableStoreFactory,
+            @NonNull final AccountID creator,
+            @NonNull final Iterator<Transaction> transactions) {
+
+        requireNonNull(readableStoreFactory);
+        requireNonNull(creator);
+        requireNonNull(transactions);
+
+        // Used for looking up payer account information.
+        final var accountStore = readableStoreFactory.createStore(ReadableAccountStore.class);
+
+        // Using the executor service, submit a task for each transaction in the event.
+        final var tasks = new ArrayList<WorkItem>(1000); // Some arbitrary number
+        while (transactions.hasNext()) {
+            // Skip platform (system) transactions.
+            final var platformTx = transactions.next();
+            if (platformTx.isSystem()) continue;
 
-    public void preHandle(@NonNull final Iterator<Transaction> itr, @NonNull final HederaState state) {
-        // Each transaction in the event will go through pre-handle using a background thread
-        // from the executor service. The Future representing that work is stored on the
-        // platform transaction. The HandleTransactionWorkflow will pull this future back
-        // out and use it to block until the pre handle work is done, if needed.
-        final ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
-        while (itr.hasNext()) {
-            final var platformTx = itr.next();
-            final var future = runner.apply(() -> {
-                try {
-                    final var metadata = securePreHandle(state, platformTx);
-                    platformTx.setMetadata(metadata);
-                } catch (final PreCheckException e) {
-                    platformTx.setMetadata(createInvalidResult(e.responseCode()));
-                }
-            });
-            futures.add(future);
+            // Submit the task to the executor service and put the resulting Future as the metadata on the transaction
+            final var future =
+                    exe.submit(() -> preHandleTransaction(creator, readableStoreFactory, accountStore, platformTx));
+            tasks.add(new WorkItem(platformTx, future));
         }
 
-        // wait until all transactions were processed before returning
-        final CompletableFuture<?>[] array = futures.toArray(new CompletableFuture<?>[0]);
-        CompletableFuture.allOf(array).join();
+        // Waits for all the background threads to complete their work and stores the resulting PreHandleResult
+        // as the transaction's metadata.
+        for (final var task : tasks) {
```

Comment:
I would have preferred to do everything in a single loop, but it is probably just a matter of personal taste.

Reply:
It could be done in a single loop using CompletableFuture and using the fluent API for that class to apply the result process on each, then just join them all at the end.  As a bonus you get a single timeout, which further simplifies things and reduces possible errors.
e.g. (note, indentation may be off; typing code in github comments is not ideal; spotless probably won't like things here; there may be bugs)
```
        var tasks = new ArrayList<CompletableFuture>(512); // generally not much benefit going over 512 due to the array of arrays ArrayList uses.
        while (transactions.hasNext()) {
            // Skip platform (system) transactions.
            final var platformTx = transactions.next();
            if (platformTx.isSystem()) continue;

            // Submit the task with its completion process to put the result as the metadata on the transaction to the executor service
            final var future =
                    CompletableFuture.runAsync(() ->
                    preHandleTransaction(creator, readableStoreFactory, accountStore, platformTx), exe)
                    .whenCompleteAsync(new PreHandleResultHandler(platformTx), exe);
            tasks.add(future);
        }
        final var config = configProvider.getConfiguration().getConfigData(PreHandleConfig.class);
        final var timeout = config.timeout();
        CompletableFuture.allOf(tasks).orTimeout(timeout).join();
        // at this point all futures are complete and each has completed the code below
        ...
   }
...
private static class PreHandleResultHandler implements BiConsumer<PreHandleResult, Throwable> {
    private final platformTransaction;

    public PreHandleResultHandler(@NonNull final Transaction platformTx) {
        platformTransaction = Objects.requiresNonNull(platformTx);
    }

    @Override
    public void accept(final PreHandleResult result, final Throwable verificationFailure) {
        if(verificationFailure == null) {
            platformTransaction.setMetadata(result);
        } else {
            switch(verificationFailure) {
                case InterruptedException e -> {
                    // The thread should only ever be interrupted during shutdown, so we can just log the error.
                    logger.error("Interrupted while waiting for a transaction to be pre handled.", e);
                    Thread.currentThread().interrupt();
                    task.platformTx.setMetadata(unknownFailure());
Member
                }
                case ExecutionException e -> {
                    logger.error("Unexpected error while pre handling a transaction!", e);
                    task.platformTx.setMetadata(unknownFailure());
                }
                case TimeoutException e -> {
                    logger.error("Timed out while waiting for a transaction to be pre handled!", e);
                    task.platformTx.setMetadata(unknownFailure());
Member
                }
            }
        }
    }
}
```

Reply:
I wanted to acknowledge before closing this conversation that I have seen this and appreciate it. I didn't see it before because this section of code was redone based on a review by @jasperpotts before I got to the comment but I really appreciate the time to write this up and I learned a few things along the way!